File: db/peek.go
package db

import (
	"database/sql"
	"fmt"
	"strings"
	"time"
)

type PeekState struct {
	SessionID  string `json:"session_id"`
	Content    string `json:"content"`
	NeedsInput bool   `json:"needs_input"`
	Reason     string `json:"reason"`
	UpdatedAt  string `json:"updated_at"`
}

func (db *DB) UpsertPeekState(sessionID, content string, needsInput bool, reason, updatedAt string) error {
	needsInputInt := 0
	if needsInput {
		needsInputInt = 1
	}
	_, err := db.conn.Exec(`
		INSERT INTO peek_state (session_id, content, needs_input, reason, updated_at)
		VALUES (?, ?, ?, ?, ?)
		ON CONFLICT(session_id) DO UPDATE SET
			content = excluded.content,
			needs_input = excluded.needs_input,
			reason = excluded.reason,
			updated_at = excluded.updated_at
	`, sessionID, content, needsInputInt, reason, updatedAt)
	if err != nil {
		return fmt.Errorf("failed to upsert peek state: %w", err)
	}
	return nil
}

func (db *DB) GetPeekState(sessionID string) (*PeekState, error) {
	var ps PeekState
	var needsInputInt int
	err := db.conn.QueryRow(`
		SELECT session_id, content, needs_input, COALESCE(reason, ''), updated_at
		FROM peek_state WHERE session_id = ?
	`, sessionID).Scan(&ps.SessionID, &ps.Content, &needsInputInt, &ps.Reason, &ps.UpdatedAt)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("failed to get peek state: %w", err)
	}
	ps.NeedsInput = needsInputInt == 1
	return &ps, nil
}

func (db *DB) ListPeekStates() ([]PeekState, error) {
	rows, err := db.conn.Query(`
		SELECT session_id, content, needs_input, COALESCE(reason, ''), updated_at
		FROM peek_state ORDER BY session_id
	`)
	if err != nil {
		return nil, fmt.Errorf("failed to list peek states: %w", err)
	}
	defer rows.Close()

	var states []PeekState
	for rows.Next() {
		var ps PeekState
		var needsInputInt int
		if err := rows.Scan(&ps.SessionID, &ps.Content, &needsInputInt, &ps.Reason, &ps.UpdatedAt); err != nil {
			return nil, fmt.Errorf("failed to scan peek state: %w", err)
		}
		ps.NeedsInput = needsInputInt == 1
		states = append(states, ps)
	}
	return states, rows.Err()
}

// PeekStatesAge returns the time since the newest updated_at in peek_state.
// Returns a very large duration if the table is empty.
func (db *DB) PeekStatesAge() (time.Duration, error) {
	var newest string
	err := db.conn.QueryRow(`SELECT MAX(updated_at) FROM peek_state`).Scan(&newest)
	if err != nil || newest == "" {
		return 24 * time.Hour, nil
	}
	t, err := time.Parse(time.RFC3339, newest)
	if err != nil {
		return 24 * time.Hour, nil
	}
	return time.Since(t), nil
}

func (db *DB) DeletePeekStatesForSessions(sessionIDs []string) error {
	if len(sessionIDs) == 0 {
		return nil
	}
	placeholders := make([]string, len(sessionIDs))
	args := make([]interface{}, len(sessionIDs))
	for i, id := range sessionIDs {
		placeholders[i] = "?"
		args[i] = id
	}
	query := fmt.Sprintf("DELETE FROM peek_state WHERE session_id IN (%s)",
		strings.Join(placeholders, ", "))
	_, err := db.conn.Exec(query, args...)
	if err != nil {
		return fmt.Errorf("failed to delete peek states: %w", err)
	}
	return nil
}

File: src/Handler/Db/Peek.hs
-- | Port of db/peek.go: cached per-session "peek" content for the statusline.
module Handler.Db.Peek
  ( PeekState(..)
  , peekStateToJson
  , upsertPeekState
  , getPeekState
  , listPeekStates
  , peekStatesAgeSeconds
  , deletePeekStatesForSessions
  ) where

import Control.Monad (unless)
import Data.Aeson (Value, object, (.=))
import Data.Text (Text)
import qualified Data.Text as T
import Data.Time.Clock (diffUTCTime, getCurrentTime)
import Data.Time.Format (defaultTimeLocale, parseTimeM)
import qualified Database.SQLite.Simple as SQL

import Handler.Db (Db(..))

data PeekState = PeekState
  { sessionId  :: Text
  , content    :: Text
  , needsInput :: Bool
  , reason     :: Text
  , updatedAt  :: Text
  } deriving (Show, Eq)

instance SQL.FromRow PeekState where
  fromRow = do
    sessionId <- SQL.field
    content <- SQL.field
    needsInputInt <- SQL.field
    reason <- SQL.field
    updatedAt <- SQL.field
    pure PeekState { needsInput = (needsInputInt :: Int) == 1, .. }

peekStateToJson :: PeekState -> Value
peekStateToJson p = object
  [ "session_id" .= p.sessionId
  , "content" .= p.content
  , "needs_input" .= p.needsInput
  , "reason" .= p.reason
  , "updated_at" .= p.updatedAt
  ]

upsertPeekState :: Db -> Text -> Text -> Bool -> Text -> Text -> IO ()
upsertPeekState db sid content needsInput reason updatedAt =
  SQL.execute db.conn
    "INSERT INTO peek_state (session_id, content, needs_input, reason, updated_at)\
    \ VALUES (?, ?, ?, ?, ?)\
    \ ON CONFLICT(session_id) DO UPDATE SET\
    \  content = excluded.content,\
    \  needs_input = excluded.needs_input,\
    \  reason = excluded.reason,\
    \  updated_at = excluded.updated_at"
    (sid, content, if needsInput then 1 else 0 :: Int, reason, updatedAt)

getPeekState :: Db -> Text -> IO (Maybe PeekState)
getPeekState db sid = do
  rows <- SQL.query db.conn
    "SELECT session_id, content, needs_input, COALESCE(reason, ''), updated_at\
    \ FROM peek_state WHERE session_id = ?"
    (SQL.Only sid)
  pure $ case rows of { (r : _) -> Just r; [] -> Nothing }

listPeekStates :: Db -> IO [PeekState]
listPeekStates db =
  SQL.query_ db.conn
    "SELECT session_id, content, needs_input, COALESCE(reason, ''), updated_at\
    \ FROM peek_state ORDER BY session_id"

-- | Seconds since the newest updated_at in peek_state; 24h if empty/unparseable.
peekStatesAgeSeconds :: Db -> IO Double
peekStatesAgeSeconds db = do
  rows <- SQL.query_ db.conn "SELECT MAX(updated_at) FROM peek_state"
  let day = 24 * 3600
  case rows of
    (SQL.Only (Just newest) : _) | not (T.null newest) ->
      case parseTimeM True defaultTimeLocale "%Y-%m-%dT%H:%M:%S%QZ" (T.unpack newest) of
        Nothing -> pure day
        Just t -> do
          now <- getCurrentTime
          pure (realToFrac (diffUTCTime now t))
    _ -> pure day

deletePeekStatesForSessions :: Db -> [Text] -> IO ()
deletePeekStatesForSessions db sids =
  unless (null sids) $ do
    let placeholders = T.intercalate ", " (replicate (length sids) "?")
    SQL.execute db.conn
      (SQL.Query ("DELETE FROM peek_state WHERE session_id IN (" <> placeholders <> ")"))
      sids
